src/Types.jl

```diff
@@ -898,63 +899,249 @@ function ensure_resolved(env::EnvCache,
+            if isfile(file)
+                registry = read_registry(file)
+                # verify_registry(registry)
+                spec = RegistrySpec(name = registry["name"],
+                                    uuid = UUID(registry["uuid"]),
+                                    url = get(registry, "repo", nothing),
+                                    path = dirname(file))
+                push!(regs, spec)
+            end
+        end
     end
-    String[joinpath(depot, "registries", r) for r in regs]
+    return regs
 end
 
-# Return paths of all registries in all depots
-function registries(; clone_default=true)::Vector{String}
-    isempty(depots()) && return String[]
+# Return `RegistrySpec`s of all registries in all depots
+function collect_registries(; clone_default=true)
+    isempty(depots()) && return RegistrySpec[]
     user_regs = abspath(depots1(), "registries")
-    # TODO: delete the following let block in Julia 1.0
-    let uncurated = joinpath(user_regs, "Uncurated"),
-        general = joinpath(user_regs, "General")
-        if ispath(uncurated) && !ispath(general)
-            mv(uncurated, general)
-            git_config_file = joinpath(general, ".git", "config")
-            cfg = read(git_config_file, String)
-            cfg = replace(cfg, r"\bUncurated\b" => "General")
-            write(git_config_file, cfg)
-        end
-    end
     if clone_default
         if !ispath(user_regs) || isempty(readdir(user_regs))
             mkpath(user_regs)
+            printpkgstyle(stdout, :Cloning, "default registries into $user_regs")
+            clone_or_cp_registries(DEFAULT_REGISTRIES)
+        end
+    end
+    return RegistrySpec[r for d in depots() for r in collect_registries(d)]
+end
+
+function populate_known_registries_with_urls!(registries::Vector{RegistrySpec})
+    known_registries = DEFAULT_REGISTRIES # TODO: Some way to add stuff here?
+    for reg in registries, known in known_registries
+        if reg.uuid !== nothing
+            if reg.uuid === known.uuid
+                reg.url = known.url
+            end
+        elseif reg.name !== nothing
+            if reg.name == known.name
+                named_regs = filter(r -> r.name == reg.name, known_registries)
+                if !all(r -> r.uuid == first(named_regs).uuid, named_regs)
+                    pkgerror("multiple registries with name `$(reg.name)`, please specify with uuid.")
+                end
+                reg.url = known.url
+            end
+        end
+    end
+end
+
+clone_or_cp_registries(regs::Vector{RegistrySpec}, depot::String=depots1()) =
+    clone_or_cp_registries(Context(), regs, depot)
+function clone_or_cp_registries(ctx::Context, regs::Vector{RegistrySpec}, depot::String=depots1())
+    populate_known_registries_with_urls!(regs)
+    # TODO Bail early on preview
+    for reg in regs
+        # clone to tmpdir first
+        tmp = mktempdir()
+        if reg.path !== nothing # copy from local source
```

Comment:
It might make sense to error out in the case where both `reg.path` and `reg.url` are `!== nothing`. Something like "Ambiguous RegistrySpec". Or maybe feedback that one was chosen over the other.